include/ginkgo/core/matrix/sparsity.hpp

```diff
@@ -0,0 +1,255 @@
+     */
+    value_type *get_value() noexcept { return value_.get_data(); }
+
+    /**
+     * @copydoc Sparsity::get_value()
+     *
+     * @note This is the constant version of the function, which can be
+     *       significantly more memory efficient than the non-constant version,
+     *       so always prefer this version.
+     */
+    const value_type *get_const_value() const noexcept
+    {
+        return value_.get_const_data();
+    }
+
+
+    /**
+     * Returns the number of elements explicitly stored in the matrix.
+     *
+     * @return the number of elements explicitly stored in the matrix
+     */
+    size_type get_num_nonzeros() const noexcept
+    {
+        return col_idxs_.get_num_elems();
+    }
+
+protected:
+    /**
+     * Creates an uninitialized Sparsity matrix of the specified size.
+     *
+     * @param exec  Executor associated to the matrix
+     * @param size  size of the matrix
+     * @param num_nonzeros  number of nonzeros
+     * @param value  the value stored by all the matrix elements.
+     */
+    Sparsity(std::shared_ptr<const Executor> exec,
+             const dim<2> &size = dim<2>{}, size_type num_nonzeros = {},
+             value_type value = one<value_type>())
+        : EnableLinOp<Sparsity>(exec, size),
+          col_idxs_(exec, num_nonzeros),
+          // avoid allocation for empty matrix
+          row_ptrs_(exec, size[0] + (size[0] > 0))
+    {
+        value_type val = value;
+        auto tmp = Array<value_type>{exec->get_master(), 1};
+        tmp.get_data()[0] = val;
+        value_ = Array<value_type>{exec, std::move(tmp)};
+    }
+
+    /**
+     * Creates a Sparsity matrix from already allocated (and initialized) row
+     * pointer and column index arrays.
+     *
+     * @tparam ColIdxsArray  type of `col_idxs` array
+     * @tparam RowPtrsArray  type of `row_ptrs` array
+     *
+     * @param exec  Executor associated to the matrix
+     * @param size  size of the matrix
+     * @param col_idxs  array of column indexes
+     * @param row_ptrs  array of row pointers
+     * @param value  value stored. (same value for all matrix elements)
+     *
+     * @note If one of `row_ptrs` or `col_idxs` is not an rvalue, not
+     *       an array of IndexType and IndexType respectively, or
+     *       is on the wrong executor, an internal copy of that array will be
+     *       created, and the original array data will not be used in the
+     *       matrix.
+     */
+    template <typename ColIdxsArray, typename RowPtrsArray>
+    Sparsity(std::shared_ptr<const Executor> exec, const dim<2> &size,
+             ColIdxsArray &&col_idxs, RowPtrsArray &&row_ptrs,
+             value_type value = 1.0)
+        : EnableLinOp<Sparsity>(exec, size),
+          col_idxs_{exec, std::forward<ColIdxsArray>(col_idxs)},
+          row_ptrs_{exec, std::forward<RowPtrsArray>(row_ptrs)}
+    {
+        value_type val = value;
+        auto tmp = Array<value_type>{exec->get_master(), 1};
```

Comment:
This should not be generated by default. If an empty `Sparsity` is generated, nothing should be allocated.